src/hotspot/share/opto/cfgnode.cpp

```diff
@@ -1779,6 +1779,70 @@ static Node* is_absolute( PhaseGVN *phase, PhiNode *phi_root, int true_path) {
   return x;
 }
 
+static Node* is_minmax(PhaseGVN* phase, PhiNode* phi_root, int true_path) {
+  assert(true_path != 0, "only diamond shape graph expected");
+
+  Node* region = phi_root->in(0);
+  IfNode* iff = region->in(1)->in(0)->as_If();
+  BoolNode* bol = iff->in(1)->as_Bool();
+  Node* cmp = bol->in(1);
+
+  // Don't transform if the branch is highly predictable
+  constexpr float infrequent_prob = PROB_UNLIKELY_MAG(2);
+  if (iff->_prob < infrequent_prob || iff->_prob > (1.0f - infrequent_prob)) {
+    return nullptr;
+  }
+
+  int false_path = 3 - true_path;
+  bool is_min = false;
+  int opcode = cmp->Opcode();
+  bool is_long = opcode == Op_CmpL;
+
+  // Ensure comparison is an integral type
+  if (opcode == Op_CmpI || opcode == Op_CmpL) {
```

Comment:
Nicer would be "bailout" ifs:
```suggestion
  if (opcode != Op_CmpI && opcode != Op_CmpL) { return nullptr; }
  bool is_long = opcode == Op_CmpL;

  // Only accept canonicalized le and lt comparisons
  int test = bol->_test._test;
  if (test != BoolTest::le && test != BoolTest::lt) { return nullptr; }
```

It is nice to have all the "bailout" checks as early as possible, and then do the "meat" of the optimization afterward.

Reply:
This also avoids nesting: nesting is difficult to read and understand. A sequence of "bailouts" is easier to read.

Reply:
I agree that this would make it easier to understand the logic, I've made this change.